gitlab/resource_gitlab_project_approval_rule.go

```diff
@@ -0,0 +1,429 @@
+				return r, nil
+			}
+		}
+	}
+
+	return nil, err
+}
+
+// createApprovalRuleID creates an ID of two parts: projectID:ruleID
+func createApprovalRuleID(ruleID int, projectID string) string {
+	return buildTwoPartID(gitlab.String(projectID), gitlab.String(strconv.Itoa(ruleID)))
+}
+
+// splitApprovalRuleID splits an approval rule into two parts, projectID:ruleID.
+// An error is returned if one occurs.
+func splitApprovalRuleID(ruleID string) (string, string, error) {
+	return parseTwoPartID(ruleID)
+}
+
+// flattenApprovalRuleUserIDs flattens a list of approval user ids into a list
+// of ints for storage in state.
+func flattenApprovalRuleUserIDs(users []*gitlab.BasicUser) []interface{} {
+	m := []interface{}{}
+	for _, user := range users {
+		m = append(m, user.ID)
+	}
+
+	return m
+}
+
+// flattenApprovalRuleGroupIDs flattens a list of approval group ids into a list
+// of ints for storage in state.
+func flattenApprovalRuleGroupIDs(groups []*gitlab.Group) []interface{} {
+	m := []interface{}{}
+	for _, group := range groups {
+		m = append(m, group.ID)
+	}
+
+	return m
+}
+
+// expandApproverIds Expands an interface into a list of ints to read from state.
+func expandApproverIds(ids interface{}, hasItems bool) []int {
+	m := []int{}
+	if hasItems {
+		for _, id := range ids.(*schema.Set).List() {
+			m = append(m, id.(int))
+		}
+	}
+
+	return m
+}
+
+// getProjectApprovalRuleUserIDs generates a list of approvers for a rule that
+// could not be found in the project or in any attached groups and returns a
+// list containing only the user IDs not found as developer <= higher in the
+// project, or developer <= in any of the rule groups.
+//
+// This method attempts to work around a bug in GitLab where there are instances
+// where the `users` attribute in the response will be empty under certain
+// conditions.
+//
+// Reference: https://gitlab.com/gitlab-org/gitlab/issues/35008
+func getProjectApprovalRuleUserIDs(client *gitlab.Client, ruleID string) ([]int, error) {
+	log.Print("[DEBUG] Looking up project approval rule User IDs as a workaround for bug in GitLab.")
+
+	userIDs := newMemberList([]int{})
+
+	rule, err := getApprovalRuleByID(client, ruleID)
+	if err != nil {
+		return userIDs.List(), err
+	}
+	users := []int{}
+
+	if len(rule.Users) > 0 {
+		for _, i := range flattenApprovalRuleUserIDs(rule.Users) {
+			users = append(users, i.(int))
+		}
+
+		return users, nil
+	}
+
+	for _, i := range flattenApprovalRuleUserIDs(rule.EligibleApprovers) {
+		users = append(users, i.(int))
+	}
+
+	projectID, _, err := splitApprovalRuleID(ruleID)
+	if err != nil {
+		return users, err
+	}
+
+	if users, err := getApproversNotInProject(client.ProjectMembers, users, projectID); err != nil {
+		return users, err
+	}
+
+	groups := []int{}
+	for _, i := range flattenApprovalRuleGroupIDs(rule.Groups) {
+		groups = append(groups, i.(int))
+	}
+
+	if users, err := getApproversNotInGroups(client.Groups, users, groups); err != nil {
```

Comment:
I tried out this PR myself, and ran into one issue. I had multiple groups each with multiple members, which applied cleanly, but then the terraform plan would always be dirty saying it was removing a user from the user_ids list. I dug into it and the error was that the `users` variable here was not getting set to the result of this function call. 

I think because `users` and `err` are already defined, you need to just use an `=` instead of `:=`. And the same thing on line 297 will probably fail for the same reason. That change fixed the plan for me.